### PR TITLE
MM-31182: Reuse the created bot user when creating a DM channel

### PR DESCRIPTION
**Pull request by agnivade**
When creating a bot, we also create a DM channel between the owner
and the bot. So we would hit a replica lag issue when trying to get the
user from the replica instance immediately after creating it.

This falls into the category of read-after-write within a single request.
We fix this by passing the already created bot user during channel creation
itself, thereby getting correctness and saving a DB query.

While here, we also make some other cosmetic improvements:

- Change userId to userID
- Rename nErr to err

https://mattermost.atlassian.net/browse/MM-31182

```release-notes
Fix a bug where creation of a bot would fail due to replica lag.
```

### CODE REVIEW

File: app/bot.go [hunk 1]
```diff
@@ -62,7 +62,11 @@ func (a *App) CreateBot(bot *model.Bot) (*model.Bot, *model.AppError) {
 	} else if ownerUser != nil {
 		// Send a message to the bot's creator to inform them that the bot needs to be added
 		// to a team and channel after it's created
-		channel, err := a.GetOrCreateDirectChannel(savedBot.UserId, bot.OwnerId)
+		botOwner, err := a.GetUser(bot.OwnerId)
+		if err != nil {
+			return nil, err
+		}
+		channel, err := a.getOrCreateDirectChannelWithUser(user, botOwner)
```
Comment (nronas) [inline app/bot.go:69]: No new users are being fetched on this 🎉

**Review by nronas — APPROVED** [review]
LGTM

File: app/channel.go [hunk 2]
```diff
@@ -382,47 +407,51 @@ func (a *App) createDirectChannel(userID string, otherUserId string) (*model.Cha
...
 		var cErr *store.ErrConflict
 		var ltErr *store.ErrLimitExceeded
 		var appErr *model.AppError
 		switch {
-		case errors.As(nErr, &invErr):
+		case errors.As(err, &invErr):
 			switch {
 			case invErr.Entity == "Channel" && invErr.Field == "DeleteAt":
-				return nil, model.NewAppError("CreateChannel", "store.sql_channel.save.archived_channel.app_error", nil, "", http.StatusBadRequest)
+				return nil, model.NewAppError("CreateDirectChannel", "store.sql_channel.save.archived_channel.app_error", nil, "", http.StatusBadRequest)
```
Comment (streamer45) [inline app/channel.go:428]: Should all these be `model.NewAppError("createDirectChannelWithUser"` instead?

Comment (agnivade) [reply app/channel.go:428]: Yeah. Let me change them.

**Review by streamer45 — APPROVED** [review]
Looks good :+1:

**Comment by agnivade** [comment]
Ahh @nronas .. your create DM optimizations now has conflicts with this PR :sob: :sob:

**Comment by nronas** [comment]
@agnivade damnnn, my last couple of days was me resolving conflicts, so many things changed around DMs. Happy to have a look for you on resolving those.

**Comment by agnivade** [comment]
Hmm .. this will be a bit hard. Both PRs are moving around significant amount of code. I will ask you for a re-review once I am done for a spot check.

**Comment by agnivade** [comment]
@nronas - Kindly check once more. Thanks :)

**Review by nronas — APPROVED** [review]
👌 Bello

**Pull request merged**